### DISCUSSION TITLE
Implement jsQuestPlus

### DISCUSSION THREAD

**Discussion by kurokida** [kurokida/jsQUEST#11] [Ideas]
Hi, @tpronk 

I almost finish to implement [jsQuestPlus](https://github.com/kurokida/jsQUEST/tree/quest_plus).

At first, could you run the two examples?
https://github.com/kurokida/jsQUEST/blob/quest_plus/Example_1stim_1psy_2resp_Watson(2017).html
https://github.com/kurokida/jsQUEST/blob/quest_plus/Example_2stim_3psy_2resp_Watson(2017).html

[Watson (2017)](https://doi.org/10.1167/17.3.10) wrote a lots of examples. For the two examples, the intensity and response of all stimuli are described. So, I demonstrated using these values, and confirmed to obtain the same estimates.

I define a JavaScript class for jsQuestPlus.
For jsQUEST (not plus) you changed the style for an ES6 module using the rollup command.
It would be great if you could do the same work again this time.

**Comment by tpronk** [comment]
Hi @kurokida,

I would love to, but I'm a bit busy at the moment to go on a hike through the Austrian Alps. Back next week!

Best, Thomas

**Reply by kurokida** [reply]
No problem at all. Have fun!

**Reply by tpronk** [reply]
On it! On my Win 10 + Firefox I get:

**Example_1stim_1psy_2resp_Watson(2017).html**  

`jsQuestPlus estimates as -20.`

**Example_2stim_3psy_2resp_Watson(2017).html**
```
Threshold estimate is -32
C0 estimate is -56
Cf estimate is 1.4000000000000001
```

**Comment by tpronk** [comment]
About making bundles for jsQuestPlus via rollup: I'd need to do a bit of restructuring; shall I go ahead or wait for future updates to `src/jsQuestPlus.js`?

**Reply by kurokida** [reply]
Thank you for your cooperation.
The code is mostly done, so you may proceed with your work.

The following is a brief description of how to use jsQuestPlus.
Detailed instructions for use will be sent to you by email.

```javascript
// At first, you have to specify psychometric functions, stimulus parameters, and psychometric parameters.
// Then, create an instance from the jsquest class.

// jsqp means jsQuestPlus. You can use any variable name you like.
const jsqp = new jsquest([func_resp0, func_resp1], [contrast_samples], [threshold_samples, slope_samples, guess, lapse_samples])

// get the most informative stimulus parameters
const stim = jsqp.getStimParams()

// update the Quest data
jsqp.update(stim, resp)

```

I would like to add a few things.

- jsQuestPlus.js depends on numeric.js only.
- I think it is unlikely that the previous jsQUEST.js and this jsQuestPlus.js will be used at the same time.

**Reply by kurokida** [reply]
I think I should change your role to make bundles for this repository, but I am not familiar with this. So, please don't hesitate to contact me.

**Reply by tpronk** [reply]
PR #12 creates the bundles. It's a small PR, because the restructuring I had in mind wasn't required after all. Might still be a good idea, so I'll post some suggestions below.

**Reply by kurokida** [reply]
Thank you very much!
I merged the PR #12 , and successfully bundled the files.
But, when I ran [the sample program](https://github.com/kurokida/jsQUEST/blob/quest_plus/Example_1stim_1psy_2resp_Watson(2017).html), I received the error message "Uncaught ReferenceError: numeric is not defined".

Since the numeric.js is included in the bundled jsQuestPlus, it is not specified in the script tag in the sample program. 

You'll see the bundled files in https://github.com/kurokida/jsQUEST/tree/quest_plus/dist

Do you know the cause of the malfunction?

**Reply by tpronk** [reply]
Ah sorry, with the current setup, `numeric` should be included with a script tag in the examples, below [this line](https://github.com/kurokida/jsQUEST/blob/f5b41b67916040d164e12487dde8cf8e08f12c01/Example_1stim_1psy_2resp_Watson(2017).html#L4)

Now, we could consider setup the examples such that they use the version of `numeric` that's bundled with jsQuestPlus. I'm not 100% how to do that, but I could plahy around a bit. Might be more elegant, since you wouldn't need to import numeric separately anymore. 

Shall I give this a go?

**Reply by kurokida** [reply]
Could you take the challenge?

Since when running jsQUEST the numeric.js does not need to be included in a script tag, I would like to have a bundle so that jsQuestPlus works without specifying the numeric.js.

Also, thank you for the explanation about the branch.
I deleted all the files unrelated to jsQuestPlus from the quest_plus branch.
In other words, we can use the main branch for jsQUEST and the quest_plus branch for jsQuestPlus.
Perhaps you can remove the description about jsQUEST from package.json and rollup.confg.js in quest_plus branch.

**Comment by tpronk** [comment]
It might be handy to remove all the code related to jsQUEST from the `quest_plus` branch?

**Reply by kurokida** [reply]
Thank you for your suggestion!

Do you mean that the files related to jsQUEST are in the main branch and the files related to jsQuestPlus are in the quest_plus branch?
I was thinking it is better that the jsQuestPlus files present in the main branch, but I'm not sure.

Do users have to switch branches every time they download the files related to jsQuestPlus?
Is it possible to include both jsQUEST and jsQuestPlus in the same release?

I'm sorry, but I'm not familiar with how github works.

**Reply by tpronk** [reply]
No worries; I gladly explain! It might be handy when other developer want to import jsQUEST and bundle it with their own releases. In the current setup there are two things that make this a bit more difficult. In `package.json`:
* The [main](https://github.com/kurokida/jsQUEST/blob/bc85cbbcddfd657e43d0f127862fd760e0b1a913/package.json#L5) property can only point at one library (either jsQUEST or jsQuestPlus)
* When installing jsQuestPlus as a module, it would currently also install interp1, even though that's only a dependency of jsQUEST (and not needed for jsQuest). See [this line](https://github.com/kurokida/jsQUEST/blob/bc85cbbcddfd657e43d0f127862fd760e0b1a913/package.json#L8)

Now, as to do a repo or a branch: a repo might be a bit easier for users in how they specify installing jsQuestPlus, but a branch would work too

**Comment by tpronk** [comment]
I'll take the challenge! :) I'll dive into it next week

**Comment by tpronk** [comment]
I managed to bundle numeric with jsQuestPlus in [my fork](https://github.com/tpronk/jsQUEST/tree/quest_plus) of the quest_plus branch. 

Here are some steps to see it in action:
* First bundle jsQuestPlus by running `rollup -c`
* Then open `Example_1stim_1psy_2resp_Watson(2017).html` in your browser
* Presently, I only modified the example to load in the bundle. You can access the staircase via `jsQuestPlus.jsquest` and numeric via `jsQuestPlus.numeric`.

Before wrapping things up, I'd like to exchange thoughts with you about how we shall name things. The `jsQuestPlus.jsquest` is a bit of a strange name. Maybe call it `jsQuestPlus.staircase`?

**Reply by kurokida** [reply]
I am very grateful for your cooperation.
However, I think the previous bundle (i.e. the way to write the numeric.js manually in a script tag) is better than the latest bundle because it would be tedious to describe each time as `jsQuestPlus.jsquest` or `jsQuestPlus.numeric`.
You have spent a lot of time on this, and I am very sorry.

What do you think about my opinion?

**Comment by tpronk** [comment]
No worries; I'm having a great time and learning a lot about modern JS in
the process.

Indeed, it's a bit tedious like that, but I can think of two ways around
it. In both approaches you'd still have the shorter syntax.

1) with the script tag we could at a little bit of script at the top, like
this:
```
jsquest = jsQuestPlus.jsquest;
numeric = jsQuestPlus.numeric;
```

2) with the ES6 import, we could write:
```
import {jsquest, numeric} from jsQuestPlus
```

How does that sound?

**Reply by kurokida** [reply]
Thank you for your suggestion.

After reading your reply, I came up with a very simple solution.
Please check [this commit](https://github.com/kurokida/jsQUEST/commit/48423b64e86a0e2955a4534efc4caef5a3b3c755).

I add the jsQuestPlus.linspace function as follows:

```javascript
static linspace(x1, x2, n){
        return numeric.linspace(x1, x2, n)
}
```

[A sample code](https://github.com/kurokida/jsQUEST/commit/48423b64e86a0e2955a4534efc4caef5a3b3c755#diff-2149c6875820c4186b499ad090b32ea4f120192fe618f382331e37e4fa000e43) seems to work properly.

I think it is better to have wrappers for functions that are expected to be used frequently by users (e.g. linspace).
I don't think there are very many of them.
Of course the users can specify a script tag for the numeric.js if they want to use a full functions.

**Reply by tpronk** [reply]
Sounds good! Let's take this route. 

1. Suggestion: Perhaps it's possible to write the wrappers slightly shorter like below? Haven't tried this out though:
`static linspace = numeric.linspace`

2. Next steps: Would you like to add the wrappers for the numeric functions you expect will be used frequently? Next, I'll merge them with mine and continue wrapping up (no pun intended :)

**Reply by kurokida** [reply]
If I can write this way, great!
I will check and get back to you later.
Thanks again for your help.

**Reply by tpronk** [reply]
My pleasure; looking forward to the next round of programming!

**Comment by kurokida** [comment]
Hi @tpronk , I successfully added the wrappers for the numeric.js. 
https://github.com/kurokida/jsQUEST/tree/quest_plus

It worked fine with the method you showed me.
I would like to change this version number to 1.0 and release it.
What do you think?

**Reply by kurokida** [reply]
I also updated the package.json and rollup.config.js using your files.
I could successfully make a bundle.

**Reply by tpronk** [reply]
Looks splendid! I guess we're ready to release indeed. I got two suggestions:
1. To adhere to semver, shall we name it 1.0.0?
2. In the README I saw a guide for using jsQuestPlus as an ES6 module. I'd read that part more as a guide on how to install the repo and bundle jsQuestPlus (into both a CommonJS and an ES6 module). Using that bundle as ES6 module is simply a matter of `
import {jsquest} from "./jsQuestPlus.module.js";`

A bit more about point 2. The nice thing about this approach is that if someone else would use jsQuestPlus as a module in their own library, and bundle that library, dependencies such as numeric are automatically bundled up into that library as well.

**Reply by kurokida** [reply]
Thank you for your response.

I have released v1.0.0 in [this way](https://github.com/kurokida/jsQUEST/releases). Users may find it a little confusing because this release page includes both jsQUEST and jsQuestPlus.
Do you have any good ideas?

I have also updated [the README](https://github.com/kurokida/jsQUEST/tree/quest_plus). I understood what you meant, but I was not sure how to change it. Could you please check if this is OK?

**Reply by tpronk** [reply]
About the release page including both jsQUEST and jsQuestPlus. Yes that could get confusing, especially if one of them gets regular updates, in which case the other gets pushed downwards. I don't really know of an elegant way of solving this in a single repo. Perhaps move to two repos after all?

About the README: Yes, that goes a long way! If it's OK with you, I'd like to do some copy-editing on the text.

**Comment by kurokida** [comment]
Thank you for your pull request.

I was not clear enough, but I had already imported the files except for the "src/jsQuestPlus.js".
Can you please confirm that I have added a comment to your pull request?

Then I created a new repository for jsQuestPlus and invited you to be a collaborator.
I'm not sure, but this will allow you to do copy-editing.
You may make changes to the repository without my permission.
Especially, please feel free to change my English if you find any part of it inappropriate.

**Reply by tpronk** [reply]
Yes, I can access the repo and directly edit the README. As a test I rephrased the opening sentence, which worked out fine

**Reply by tpronk** [reply]
Rewriting completed! What do you think of it?

NB. I left a little section open on how to import jsQuestPlus as a UMD module. https://github.com/kurokida/jsQuestPlus/blob/main/README.md#as-a-umd-module

**Reply by kurokida** [reply]
Thanks to you, the README is now very clear.

I added the demonstration section. What do you think of it?

I also invite you to be a collaborator on the jsQUEST repository.
Best regards,

**Reply by tpronk** [reply]
Looks good! Got only one suggestion: for each demo, maybe have a link to the running example and to the source code in the repo?

Next up I'll do some rewriting with the jsQUEST README, if that's OK

**Reply by kurokida** [reply]
Thanks. I followed your suggestion and updated the README a bit.

**Reply by tpronk** [reply]
I just wondered: since jsQuestPlus is a more generic version of jsQUEST, it should be able to everything that jsQUEST can do (and more) right? In that case, is there any reason to recommend using jsQUEST instead of jsQuestPlus?
